fix(components): [form] improve types

ModyQyW opened the pull request:
Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

This PR was started as a discussion at https://github.com/element-plus/element-plus/discussions/12373.

In element-plus@2.3.3, we have `export declare type FormRules = Partial<Record<string, Arrayable<FormItemRule>>>;`, which means the keys of the rules object can be any string and it may causes unexpected errors. An example shows below.

```typescript
// use reactive from vue
const form = reactive({
  id: 0,
  name: '',
});
// or use ref from vue
const form = ref({
  id: 0,
  name: '',
});

// the keys of the rules object can be any string
// even not in the `form` model
const formRules: FormRules = {
  id: { ... },
  name: { ... },
  // no ts error
  code: { ... },
  ...,
};
```

This PR adds an optional generic to the `FormRules` type and brings better type assertion if someone needs it.

```typescript
const formRules: FormRules<typeof form> = {
  id: { ... },
  name: { ... },
  // ts error
  // Type '{ id: {}; name: {}; code: {}; }' is not assignable to type 'Partial<Record<"id" | "name", Arrayable<FormItemRule>>>'.
  // Object literal may only specify known properties, and 'code' does not exist in type 'Partial<Record<"id" | "name", Arrayable<FormItemRule>>>'.ts(2322)
  code: { ... },
  ...,
};
```

Previous codes still work because the generic is optional.

```typescript
const formRules: FormRules = {
  id: { ... },
  name: { ... },
  // no ts error
  code: { ... },
  ...,
};
```

ModyQyW committed:
fix(components): [form] improve types

tolking commented:
It's a good idea, but it doesn't work in all cases right now

eg.

```
const form = ref({
  id: 0,
  name: '',
  obj: { a: '' },
  arr: [{ value: '' }],
  //...
});
const formRules = {
  id: {},
  name: {},
  'obj.a': {},
  'arr.0.value': {},
  //...
}
```

ModyQyW commented:
I think writing `obj.a` is sensible. I will try to cover this case.

However, I think writing`arr.0.value` in `formRules` is kind of weird. It is hard to set rules in types for every array element, and it is easier and better to bind rules on ElFormItem. Should it be supported?

ModyQyW commented:
@tolking Hi. I have just updated the type. Can you please review it? Thanks!

ModyQyW committed:
fix(components): [form] improve FormRules type, support object and array

tolking commented:
I think we should support all possible cases

ModyQyW commented:
> It's a good idea, but it doesn't work in all cases right now
> 
> eg.
> 
> ```
> const form = ref({
>   id: 0,
>   name: '',
>   obj: { a: '' },
>   arr: [{ value: '' }],
>   //...
> });
> const formRules = {
>   id: {},
>   name: {},
>   'obj.a': {},
>   'arr.0.value': {},
>   //...
> }
> ```

> I think we should support all possible cases

Hi @tolking. The latest submission should cover the cases you listed. Can you please review it and give me some advice/suggestions? Thanks! 🙏

<img width="323" alt="image" src="https://github.com/element-plus/element-plus/assets/31850793/c67068cc-dada-4e05-9ea9-5d5ea0544c2b">

<img width="858" alt="image" src="https://github.com/element-plus/element-plus/assets/31850793/2754c028-6aa7-46d1-a646-f3bdfed1bb69">

tolking commented on packages/components/form/src/types.ts:
```diff
@@ -19,7 +20,55 @@ export type FormLabelWidthContext = ReturnType<typeof useFormLabelWidth>
+  : T extends object
+  ? { [K in keyof T]-?: Join<K, Leaves<T[K], Prev[D]>> }[keyof T]
+  : ''
+export type FormRules<
+  T extends MaybeRef<Record<string, any> | string> = string
+> = Partial<
+  Record<
+    UnwrapRef<T> extends string ? UnwrapRef<T> : Leaves<UnwrapRef<T>>,
+    Arrayable<FormItemRule>
+  >
+>
```
IMO, Perhaps we should only support simple types, using MaybeRef seems redundant. what do you think?

example

```ts
interface Form {
  a: number
}

const form = ref<Form>() // or reactive<Form>()

const formRules: FormRules<Form> = {
  a: { ... },
}
```

tolking commented on packages/components/form/src/types.ts:
```diff
@@ -19,7 +20,55 @@ export type FormLabelWidthContext = ReturnType<typeof useFormLabelWidth>
+  : T extends object
+  ? { [K in keyof T]-?: Join<K, Leaves<T[K], Prev[D]>> }[keyof T]
+  : ''
+export type FormRules<
```
It would be even better if you added some explanations to the type.

[link](https://github.com/element-plus/element-plus/blob/dev/packages/utils/vue/props/types.ts)

ModyQyW committed:
fix(components): [form] improve FormRules type with comment

ModyQyW commented:
@tolking Hi. I have just updated the types, added comments, and fix some type errors.

> Now TS has optimized the processing of recurs, it no longer needs to be written like this.

Thanks! I have removed it.

> IMO, Perhaps we should only support simple types, using `MaybeRef` seems redundant. what do you think?

I think it is a bad idea. It will increase the type burden on the user side - to use this feature, users will have to write the type by hand and could not easily resort to TypeScript type derivation.

> It would be even better if you added some description to the type.

Thanks! I have added some.

tolking commented on packages/components/form/src/types.ts:
```diff
@@ -19,7 +21,82 @@ export type FormLabelWidthContext = ReturnType<typeof useFormLabelWidth>
+type Path<T> = T extends ReadonlyArray<infer V>
+  ? IsTuple<T> extends true
+    ? {
+        [K in TupleKey<T>]-?: PathImpl<K & string, T[K]>
+      }[TupleKey<T>] // tuple
+    : PathImpl<ArrayKey, V> // array
+  : {
+      [K in keyof T]-?: PathImpl<K & string, T[K]>
```
> PathImpl<K extends string | number, V>

Why use `K & string`? For an object, the number can also be used as a key.

ModyQyW replied:
Yes you are right. I just missed that.

tolking commented on packages/components/form/src/types.ts:
```diff
@@ -19,7 +21,82 @@ export type FormLabelWidthContext = ReturnType<typeof useFormLabelWidth>
+ * 通过一个类型收集所有路径的类型
+ *
+ * @example
+ * FieldPath<{ a: number; b: string; c: { d: number; e: string }; f: [{ value: string }]; g: { value: string }[] }> => 'a' | 'b' | 'c.d' | 'c.e' | 'f.0' | 'f.0.value' | 'g.number' | 'g.number.value'
```
I found that the generated type does not contain `'c' ...`, but some components in the component library accept an object, such as ElSelect

ModyQyW replied:
That's weird. After resolving the issue above, the type seems to contain `c` already in my side. I also update the jsdoc. Can you please check that again?

<img width="814" alt="image" src="https://github.com/element-plus/element-plus/assets/31850793/167b5e88-2735-4081-9e9c-de1783c3ee80">

ModyQyW committed:
fix(components): [form] allow number as key in FormRules type

ModyQyW committed:
chore(components): [form] fix FormRules type example

ModyQyW committed:
chore: remove extra import

tolking reviewed (approved):
Can you update the types in the examples?

ModyQyW commented:
> Can you update the types in the examples?

Do you mean `element-plus/docs/examples/form`? Of course, I can.

Before I do this, I want to confirm:

1. Should I add a new part after [Number Validate](https://element-plus.org/en-US/component/form.html#number-validate) or somewhere else? Or should I just update all validation examples' types?

2. Which way do you prefer? Use `interface Form` in the code as a recommendation and only mention `typeof form` in the `:::demo` description, Or reverse it?

tolking commented:
[custom-validation](https://github.com/element-plus/element-plus/blob/dev/docs/examples/form/custom-validation.vue#L82) and [validation](https://github.com/element-plus/element-plus/blob/dev/docs/examples/form/validation.vue#L100)

I recommend use the `interface Form`

ModyQyW committed:
chore: update form/custom-validation and form/validation examples

ModyQyW commented:
> [custom-validation](https://github.com/element-plus/element-plus/blob/dev/docs/examples/form/custom-validation.vue?rgh-link-date=2023-05-25T05%3A11%3A03Z#L82) and [validation](https://github.com/element-plus/element-plus/blob/dev/docs/examples/form/validation.vue?rgh-link-date=2023-05-25T05%3A11%3A03Z#L100)
> 
> I recommend use the `interface Form`

I just push commits about examples. If I misunderstand something, please tell me and I will correct them.

ModyQyW commented:
Just rebased.

btea commented on docs/examples/form/custom-validation.vue:
```diff
@@ -73,13 +73,18 @@ const validatePass2 = (rule: any, value: any, callback: any) => {
+  checkPass: string
+  age: string
+}
+const ruleForm = reactive<RuleForm>({
```
'RuleForm' can probably be removed.

btea commented on docs/examples/form/custom-validation.vue:
```diff
@@ -73,13 +73,18 @@ const validatePass2 = (rule: any, value: any, callback: any) => {
 })
 
-const rules = reactive<FormRules>({
+const rules = reactive<FormRules<RuleForm>>({
```
```suggestion
const rules = reactive<FormRules<typeof ruleForm>>({
```

ModyQyW committed:
chore(components): [form] update example

ModyQyW commented:
@btea Hi. I just update the example. If I misunderstand something, please tell me and I will correct them. Thanks. 🙏

btea commented:
Your understanding is absolutely correct. 👍

btea reviewed (approved):
LGTM
